### DISCUSSION TITLE
Should I use a Riverpod provider to hold the model data for a complex form?

### DISCUSSION THREAD

**Discussion by davoutuk** [rrousselGit/riverpod#2268] [General]
I've been using RiverPod for a while now with Flutter, but I'm not sure what the best practice is for how to use RiverPod with a complex form.

Assume that I have a complex form that mas multiple text inputs., drop down selectors and radio buttons.    For this I am using a Flutter '<form>' widget to manager validation and save actions.  I have a model class that holds the data that is used to initiate the controls and capture the save action results. 

Should I create a distinct RiverPod StateNotifier provider that uses the form model as its state, or should I keep the model data local to the form widget? 

The downside of using a RiverPod StateProvider provider is that I have to create a 'setXXX' method in the StateNotifier' for each data entry widget n the form. That way when the form save action is called each data entry widget can call a related method in the StateNotifier to update its property into the model 

Any suggestions?

**Comment by rrousselGit** [comment]
For forms, I would recommend keeping the state local to your widgets (possibly using hooks if that's your thing). And put the state inside providers only on submit after it has been validated

**Reply by SiddiquiMonis401** [reply]
And what about other states which does not involve forms ?? Is it good to kept them in Notifierprovider

**Reply by njwandroid** [reply]
@rrousselGit i like hooks... but wouldn't using a provider of some sort help for automated testing as well as modularity and keeping the widget smaller? It feels nice to keep as much out of the widget as you can... at least  it feels like if there's say 7-12+ input fields there should be a better way than 7-12 useState(String) variables... useState(FromWrapperClass()) i suppose then?

**Reply by ivankasalo** [reply]
@njwandroid I completely agree with you and still I am not sure what would be the best approach when dealing with complex forms.
